### PR TITLE
Ear Plugin: application descriptors for Jakarta EE 12+

### PR DESCRIPTION
**Pull request by p-schneider**
It's better to assume the schema for ear application descriptor files doesn't change in upcoming versions than generating a application xml file that most probably isn't correct.
If a compatibility issue raises in a future Jakarta EE Release, the generation for the descriptor could then be fixed in an upcoming Gradle release and could be worked around by using a custom xml descriptor until then.

This is a followup to #34342 / #34343

### Context
See comments at https://github.com/gradle/gradle/issues/34342#issuecomment-3103329757

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md).
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team.
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective.
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic.
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes.
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`.
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by p-schneider** [comment]
@ljacomet this is a followup on what you commented last month on #34342.

File: platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/descriptor/internal/DefaultDeploymentDescriptor.java [hunk 1]
```diff
@@ -437,10 +437,8 @@ private Object nodeNameFor(String name) {
             return new QName("http://java.sun.com/xml/ns/javaee", name);
         } else if ("7".equals(version) || "8".equals(version)) {
             return new QName("http://xmlns.jcp.org/xml/ns/javaee", name);
-        } else if ("9".equals(version) || "10".equals(version) || "11".equals(version)) {
-            return new QName("https://jakarta.ee/xml/ns/jakartaee", name);
         } else {
-            return new QName(name);
+            return new QName("https://jakarta.ee/xml/ns/jakartaee", name);
```
Comment (wheeleme) [inline platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/descriptor/internal/DefaultDeploymentDescriptor.java:447]: ❌ We should keep the empty branch. A solution would be to verify the value we receive is an Integer. In that case, use `jakarta.ee`, otherwise use `QName(name)`. A similar change would be done for the other method.

**Review by wheeleme — DISMISSED** [review]
Thank you for your contribution. Please take a look at the comments in the file and resubmit the changes.

File: platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/descriptor/internal/DefaultDeploymentDescriptor.java (hunk 1 above)
Comment (p-schneider) [reply platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/descriptor/internal/DefaultDeploymentDescriptor.java:447]: I've pushed changes to the implementation to match any version number from `9` to `99` as Jakarta EE, and otherwise use the fallback without an XML namespace.

**Comment by ov7a** [comment]
@p-schneider PTAL at failing tests in EarPluginIntegrationTest
* https://ge.gradle.org/scans/tests?search.names=gitCommitId,PartOf&search.tags=CI&search.timeZoneId=Atlantic%2FReykjavik&search.values=6e6b0025bc5b7cfe09353b4f34e720a5e4a61296,ReadyforNightly#

**Comment by p-schneider** [comment]
> @p-schneider PTAL at failing tests in EarPluginIntegrationTest

I added a null check (`version != null`) before calling `JAKARTA_VERSION_PATTERN.matcher(version).matches()` which fixes the issue. (Running [EarPluginIntegrationTest](https://github.com/gradle/gradle/blob/master/platforms/jvm/ear/src/integTest/groovy/org/gradle/plugins/ear/EarPluginIntegrationTest.groovy) locally confirmed that this was the relevant change needed.)

Should I also add a test for version=null to the unit tests in [DefaultDeploymentDescriptorTest](https://github.com/gradle/gradle/blob/master/platforms/jvm/ear/src/test/groovy/org/gradle/plugins/ear/descriptor/internal/DefaultDeploymentDescriptorTest.groovy)?

**Comment by p-schneider** [comment]
> @p-schneider PTAL at failing tests in EarPluginIntegrationTest

In my latest rebased/amended commit I added a null check (`version != null`) before calling `JAKARTA_VERSION_PATTERN.matcher(version).matches()` which fixes the issue. (Running [EarPluginIntegrationTest](https://github.com/gradle/gradle/blob/master/platforms/jvm/ear/src/integTest/groovy/org/gradle/plugins/ear/EarPluginIntegrationTest.groovy) locally confirmed that this was the relevant change needed.)

@ov7a So you can re-review and add this to the merge queue again.

Should I also add a test for version=null to the unit tests in [DefaultDeploymentDescriptorTest](https://github.com/gradle/gradle/blob/master/platforms/jvm/ear/src/test/groovy/org/gradle/plugins/ear/descriptor/internal/DefaultDeploymentDescriptorTest.groovy)?

**Comment by ov7a** [comment]
@bot-gradle test this

**Comment by ov7a** [comment]
> So you can re-review and add this to the merge queue again.

LGTM!  I've triggered the build.

> Should I also add a test for version=null to the unit tests in [DefaultDeploymentDescriptorTest](https://github.com/gradle/gradle/blob/master/platforms/jvm/ear/src/test/groovy/org/gradle/plugins/ear/descriptor/internal/DefaultDeploymentDescriptorTest.groovy)?

That would be nice, thank you!

**Comment by p-schneider** [comment]
> > Should I also add a test for version=null to the unit tests in [DefaultDeploymentDescriptorTest](https://github.com/gradle/gradle/blob/master/platforms/jvm/ear/src/test/groovy/org/gradle/plugins/ear/descriptor/internal/DefaultDeploymentDescriptorTest.groovy)?
> 
> That would be nice, thank you!

@ov7a 
I added Unit tests for `version = null` and made `version="null"` disappear from the xml output.
After all `null` isn't a useful value, but at least it doesn't crash and instead omits the version parameter in the xml output.
So the PR should be good to go.
